chore: switch to pnpm

mcmxcdev opened the pull request:
## Description

Well, here we go again...
I think I definitely got further than back with Ferdi, I was able to build all the linux output with `pnpm build`

I got stopped by a runtime error when attempting to launch the AppImage:

![Screenshot from 2022-04-22 21-34-06](https://user-images.githubusercontent.com/16797721/164782973-fd3892c8-94e0-4d60-8a5d-1161cdfcc38c.png)

I saw that there is a new major version of electron-builder, so it might make sense to upgrade that one, but who knows if it fixes this.

## Changes
- upgrade `sqlite3` to fix error with `build` npm script
- add missing reported peer dependency `@babel/core`
- add `pnpm-lock.yaml`
- remove `package-lock.json`

nathanaelhoun commented on scripts/build-unix.sh:
```diff
@@ -92,38 +92,31 @@ fi
+# Check pnpm version of the recipes submodule
+EXPECTED_RECIPES_PNPM_VERSION=$(node -p 'require("./recipes/package.json").engines.pnpm')
+if [[ "$EXPECTED_PNPM_VERSION" != "$EXPECTED_RECIPES_PNPM_VERSION" ]]; then
+  fail_with_docs "You are not running the expected version of pnpm between the main repo and the recipes submodule!
+    expected: [$EXPECTED_RECIPES_PNPM_VERSION]
+    actual  : [$EXPECTED_PNPM_VERSION]"
```
```suggestion
  fail_with_docs "The expected versions of pnpm are not the same in the main repo and in the recipes submodule, please sync them.
    expected in recipes  : [$EXPECTED_RECIPES_PNPM_VERSION]
    expected in main repo: [$EXPECTED_RECIPES_PNPM_VERSION]
    actual               : [$EXPECTED_PNPM_VERSION]"
```

nathanaelhoun commented on scripts/build-windows.ps1:
```diff
@@ -118,44 +118,28 @@ if ([double]$ACTUAL_MSVST_VERSION -ne [double]$EXPECTED_MSVST_VERSION) {
+# Check pnpm version of the recipes submodule
+$EXPECTED_RECIPES_PNPM_VERSION = (Get-Content .\recipes\package.json | ConvertFrom-Json).engines.pnpm
+if ($ACTUAL_PNPM_VERSION -ne $EXPECTED_RECIPES_PNPM_VERSION) {
+  fail_with_docs "You are not running the expected version of pnpm between the main repo and the recipes submodule!
+    expected: [$EXPECTED_RECIPES_PNPM_VERSION]
+    actual  : [$ACTUAL_PNPM_VERSION]"
```
```suggestion
 fail_with_docs "The expected versions of pnpm are not the same in the main repo and in the recipes submodule, please sync them.
    expected in recipes  : [$EXPECTED_RECIPES_PNPM_VERSION]
    expected in main repo: [$EXPECTED_RECIPES_PNPM_VERSION]
    actual               : [$EXPECTED_PNPM_VERSION]"
```

vraravam commented:
I was able to build the app correctly on macos. But, when I attempted to run it, it came back with the followin error:
<img width="264" alt="Screenshot 2022-05-01 at 1 26 08 AM" src="https://user-images.githubusercontent.com/69629/166134925-767ecfc1-8720-434f-b467-eaf5475f768a.png">

The error stack is:
```
Uncaught Exception:
Error: Cannot find module 'jsonfile/utils'
Require stack:

- /Applications/Ferdium.app/Contents/Resources/app.asar/node_modules/fs-extra/lib/json/output-json.js
- /Applications/Ferdium.app/Contents/Resources/app.asar/node_modules/fs-extra/lib/json/index.js
- /Applications/Ferdium.app/Contents/Resources/app.asar/node_modules/fs-extra/lib/index.js
- /Applications/Ferdium.app/Contents/Resources/app.asar/index.js

-
at Module._resolveFilename (node:internal/modules/cjs/loader:940:15)
at Function.n._resolveFilename (node:electron/js2c/browser_init:245:1105)
at Module._load (node:internal/modules/cjs/loader:785:27)
at Function.c._load (node:electron/js2c/asar_bundle:5:13343)
at Module.require (node:internal/modules/cjs/loader:1012:19)
at require (node:internal/modules/cjs/helpers:102:18)
at Object.<anonymous> (/Applications/Ferdium.app/Contents/Resources/app.asar/node_modules/fs-extra/lib/json/output-json.js:3:23)
at Module._compile (node:internal/modules/cjs/loader:1116:14)
at Object.Module._extensions..js (node:internal/modules/cjs/loader:1169:10)
at Module.load (node:internal/modules/cjs/loader:988:32)
```
@mcmxcdev - would you be able to take this forward? I feel we are very close to solving the problem. It might be that we need to add some new flags into `.npmrc` for this to work?

vraravam commented:
If I add the `jsonfile` package to the `pnpm` lock file, then the app process starts up - but no window shows up. This has been tested on macos.

santhosh-chinnasamy commented:
while running `CLEAN=true ./scripts/build-unix.sh` I got few errors

`.npmrc` config

```npmrc
save-exact=true
engine-strict=true
unsafe-perm=true
frozen-lockfile=false
# prefer-frozen-lockfile=false

auto-install-peers=true
# legacy-peer-deps=true
public-hoist-pattern=*
package-import-method=copy
shamefully-hoist=true
node-linker=hoisted
```

![image](https://user-images.githubusercontent.com/20743451/166413508-63bc8078-06a9-4e92-b40b-46f38d0021cd.png)

![image](https://user-images.githubusercontent.com/20743451/166413637-a7196731-79cf-43ac-bb3c-66df912ea741.png)

![image](https://user-images.githubusercontent.com/20743451/166413693-5b3ed606-c174-42f9-a7e6-37aed80493db.png)

vraravam commented:
@santhosh-chinnasamy - please see my last commit - maybe you can revert that and proceed?

santhosh-chinnasamy commented:
@vraravam I am unable to find any solution to fix this.

Alphrag commented:
After a rebase of the branch from develop and a revert of `4dc75eb` , I have tried to build it on w10 and I'm currently with the following error (note that it happens later than it has ever done before):
```
src\components\ui\AppLoader\index.tsx(77,1): error TS2742: The inferred type of 'default' cannot be named without a reference to '.pnpm/@types+react@17.0.44/node_modules/@types/react'. This is likely not portable. A type annotation is necessary.
src\components\ui\badge\index.tsx(70,14): error TS2742: The inferred type of 'Badge' cannot be named without a reference to '.pnpm/@types+react@17.0.44/node_modules/@types/react'. This is likely not portable. A type annotation is necessary.
src\components\ui\badge\ProBadge.tsx(63,14): error TS2742: The inferred type of 'ProBadge' cannot be named without a reference to '.pnpm/@types+react@17.0.44/node_modules/@types/react'. This is likely not portable. A type annotation is necessary.
src\components\ui\button\index.tsx(262,14): error TS2742: The inferred type of 'Button' cannot be named without a reference to '.pnpm/@types+react@17.0.44/node_modules/@types/react'. This is likely not portable. A type annotation is necessary.
src\components\ui\error\index.tsx(20,14): error TS2742: The inferred type of 'Error' cannot be named without a reference to '.pnpm/@types+react@17.0.44/node_modules/@types/react'. This is likely not portable. A type annotation is necessary.
src\components\ui\icon\index.tsx(45,14): error TS2742: The inferred type of 'Icon' cannot be named without a reference to '.pnpm/@types+react@17.0.44/node_modules/@types/react'. This is likely not portable. A src\components\ui\Loader.tsx(44,1): error TS2742: The inferred type of 'default' cannot be named without a reference to '.pnpm/mobx-react@6.3.1_39d463ce7769f1a2935ab17929e2e668/node_modules/mobx-react'. This is likely not portable. A type annotation is necessary.
src\components\ui\loader\index.tsx(44,14): error TS2742: The inferred type of 'Loader' cannot be named without a reference to '.pnpm/@types+react@17.0.44/node_modules/@types/react'. This is likely not portable. A type annotation is necessary.
src\components\ui\Modal\index.tsx(66,1): error TS2742: The inferred type of 'default' cannot be named without a reference to '.pnpm/@types+react@17.0.44/node_modules/@types/react'. This is likely not portable. A type annotation is necessary.
src\components\ui\select\index.tsx(461,14): error TS2742: The inferred type of 'Select' cannot be named without a reference to '.pnpm/@types+react@17.0.44/node_modules/@types/react'. This is likely not portable. A type annotation is necessary.
src\components\ui\textarea\index.tsx(124,14): error TS2742: The inferred type of 'Textarea' cannot be named without a reference to '.pnpm/@types+react@17.0.44/node_modules/@types/react'. This is likely not portable. A type annotation is necessary.
src\components\ui\toggle\index.tsx(125,14): error TS2742: The inferred type of 'Toggle' cannot be named without a reference to '.pnpm/@types+react@17.0.44/node_modules/@types/react'. This is likely not portable. A type annotation is necessary.
src\components\ui\wrapper\index.tsx(36,14): error TS2742: The inferred type of 'Wrapper' cannot be named without a reference to '.pnpm/@types+react@17.0.44/node_modules/@types/react'. This is likely not portable. A type annotation is necessary.
src\features\basicAuth\store.ts(12,14): error TS2742: The inferred type of 'state' cannot be named without a reference to '.pnpm/mobx@5.15.7/node_modules/mobx'. This is likely not portable. A type annotation is necessary.
src\features\publishDebugInfo\store.ts(7,14): error TS2742: The inferred type of 'state' cannot be named without a reference to '.pnpm/mobx@5.15.7/node_modules/mobx'. This is likely not portable. A type annotation is necessary.
src\features\quickSwitch\store.ts(7,14): error TS2742: The inferred type of 'state' cannot be named without a reference to '.pnpm/mobx@5.15.7/node_modules/mobx'. This is likely not portable. A type annotation is necessary.
src\features\serviceProxy\index.ts(6,14): error TS2742: The inferred type of 'config' cannot be named without a reference to '.pnpm/mobx@5.15.7/node_modules/mobx'. This is likely not portable. A type annotation is necessary.
src\features\workspaces\components\WorkspaceItem.tsx(43,1): error TS2742: The inferred type of 'default' cannot be named without a reference to '.pnpm/@types+react@17.0.44/node_modules/@types/react'. This is likely not portable. A type annotation is necessary.
src\features\workspaces\components\WorkspaceServiceListItem.tsx(64,1): error TS2742: The inferred type of 'default' cannot be named without a reference to '.pnpm/@types+react@17.0.44/node_modules/@types/react'. This is likely not portable. A type annotation is necessary.
src\features\workspaces\containers\EditWorkspaceScreen.tsx(59,1): error TS2742: The inferred type of 'default' cannot be named without a reference to '.pnpm/mobx-react@6.3.1_39d463ce7769f1a2935ab17929e2e668/node_modules/mobx-react'. This is likely not portable. A type annotation is necessary.
src\features\workspaces\containers\WorkspacesScreen.tsx(39,1): error TS2742: The inferred type of 'default' cannot be named without a reference to '.pnpm/mobx-react@6.3.1_39d463ce7769f1a2935ab17929e2e668/node_modules/mobx-react'. This is likely not portable. A type annotation is necessary.
src\I18n.tsx(40,1): error TS2742: The inferred type of 'default' cannot be named without a reference to '.pnpm/mobx-react@6.3.1_39d463ce7769f1a2935ab17929e2e668/node_modules/mobx-react'. This is likely not portable. A type annotation is necessary.
src\routes.tsx(84,1): error TS2742: The inferred type of 'default' cannot be named without a reference to '.pnpm/mobx-react@6.3.1_39d463ce7769f1a2935ab17929e2e668/node_modules/mobx-react'. This is likely not portable. A type annotation is necessary.
TypeScript: 26 emit errors
TypeScript: emit failed
[23:41:14] Error: TypeScript: Compilation failed
    at Output.mightFinish (D:\Ferdium\ferdium-app\node_modules\.pnpm\gulp-typescript@6.0.0-alpha.1_typescript@4.6.4\node_modules\gulp-typescript\release\output.js:131:43)
    at D:\Ferdium\ferdium-app\node_modules\.pnpm\gulp-typescript@6.0.0-alpha.1_typescript@4.6.4\node_modules\gulp-typescript\release\output.js:44:22
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
``` 
![pnpm_build_error-2022-05-08](https://user-images.githubusercontent.com/34252790/167274821-a0b20cd6-4465-44aa-97bc-85751b18eec3.png)

mcmxcdev committed:
Switch from 'npm' to 'pnpm' (WIP: packaged app doesnt start)

- replace 'package-lock.json' with 'pnpm-lock.yaml'
- lock 'node' and 'pnpm' to specific versions

Co-authored-by: Nathanaël Houn <contact@nathanaelhoun.fr>
Co-authored-by: Vijay A <vraravam@users.noreply.github.com>

vraravam commented:
Superceded by https://github.com/ferdium/ferdium-app/pull/704
